File: php/mapper.php
#!/usr/bin/env php
<?php

const DATE = 0;
const ZIP_CODE = 2;
const FIRST_STREET = 6;
const SECOND_STREET = 7;
const THIRD_STREET = 8;
const PEDESTRIANS_INJURED = 11;
const PEDESTRIANS_KILLED = 12;
const CYCLIST_INJURED = 13;
const CYCLIST_KILLED = 14;
const MOTORIST_INJURED = 15;
const MOTORIST_KILLED = 16;

function getVictimType($index) {
    switch($index) {
        case PEDESTRIANS_INJURED:
        case PEDESTRIANS_KILLED:
            return "PEDESTRIAN";
        case CYCLIST_INJURED:
        case CYCLIST_KILLED:
            return "CYCLIST";
        case MOTORIST_INJURED:
        case MOTORIST_KILLED:
            return "MOTORIST";
        default:
            echo "[ERROR] VICTIM TYPE UNKNOWN: " . $index;
            exit(111);
    }
}

function getInjuriesType($index) {
    switch ($index) {
        case PEDESTRIANS_INJURED:
        case CYCLIST_INJURED:
        case MOTORIST_INJURED:
            return "INJURED";
        case PEDESTRIANS_KILLED:
        case CYCLIST_KILLED:
        case MOTORIST_KILLED:
            return "KILLED";
        default:
            echo "[ERROR] INJURIES TYPE UNKNOWN: " . $index;
            exit(111);
    }
}

function mapRow($row) {
    $date = explode("/", $row[DATE]);
    if(count($date) != 3) {
        return;
    }

    $year = (int)($date[2]);
    $zipCode = $row[ZIP_CODE];

    if ($year <= 2012 || $zipCode === "") {
        return;
    }

    for ($streetIndex = FIRST_STREET; $streetIndex <= THIRD_STREET; $streetIndex++) {
        $street = $row[$streetIndex];
        if($street === "") {
            continue;
        }
        for ($numberIndex = PEDESTRIANS_INJURED; $numberIndex <= MOTORIST_KILLED; $numberIndex++) {
            $victimType = getVictimType($numberIndex);
            $injuriesType = getInjuriesType($numberIndex);
            $number = $row[$numberIndex];
            printf("%s|%s|%s|%s\t%s\n", $street, $zipCode, $victimType, $injuriesType, $number);
        }
    }
}

fgets(STDIN);
while(($line = fgets(STDIN)) !== false) {
    $tokens = explode(",", $line);
    mapRow($tokens);
}

File: python/mapper.py
#!/usr/bin/env python3

import sys

DATE = 0
ZIP_CODE = 2
FIRST_STREET = 6
SECOND_STREET = 7
THIRD_STREET = 8
PEDESTRIANS_INJURED = 11
PEDESTRIANS_KILLED = 12
CYCLIST_INJURED = 13
CYCLIST_KILLED = 14
MOTORIST_INJURED = 15
MOTORIST_KILLED = 16


def get_victim_type(index):
    return {
        PEDESTRIANS_INJURED: "PEDESTRIAN",
        PEDESTRIANS_KILLED: "PEDESTRIAN",
        CYCLIST_INJURED: "CYCLIST",
        CYCLIST_KILLED: "CYCLIST",
        MOTORIST_INJURED: "MOTORIST",
        MOTORIST_KILLED: "MOTORIST",
    }[index]


def get_injuries_type(index):
    return {
        PEDESTRIANS_INJURED: "INJURED",
        CYCLIST_INJURED: "INJURED",
        MOTORIST_INJURED: "INJURED",
        PEDESTRIANS_KILLED: "KILLED",
        MOTORIST_KILLED: "KILLED",
        CYCLIST_KILLED: "KILLED",
    }[index]


def map_row(row):
    date = row[DATE].split("/")
    if len(date) != 3:
        return

    year = int(date[2])
    zip_code = row[ZIP_CODE].strip()

    if year <= 2012 or zip_code == "":
        return

    for street_index in range(FIRST_STREET, THIRD_STREET + 1):
        street = row[street_index]
        if street == "":
            continue

        for number_index in range(PEDESTRIANS_INJURED, MOTORIST_KILLED + 1):
            victim_type = get_victim_type(number_index)
            injuries_type = get_injuries_type(number_index)
            number = int(row[number_index])
            print(f"{street}|{zip_code}|{victim_type}|{injuries_type}\t{number}")


if __name__ == "__main__":
    sys.stdin.readline()
    for line in sys.stdin:
        tokens = line.strip().split(",")
        if len(tokens) != 28:
            continue
        map_row(tokens)
